programs/drift_vaults/src/state/vault.rs

```diff
@@ -130,45 +148,209 @@ impl Vault {
                 .safe_div(depositor_equity.safe_sub(management_fee_payment)?)?
                 .cast()?;
 
-            let new_total_shares = self
+            let new_total_shares = vault
                 .total_shares
                 .safe_mul(new_total_shares_factor.cast()?)?
                 .safe_div(PERCENTAGE_PRECISION)?
-                .max(self.user_shares);
+                .max(vault.user_shares);
 
-            if management_fee_payment == 0 || self.total_shares == new_total_shares {
-                // time delta wasnt large enough to pay any management fee
+            if management_fee_payment == 0 || vault.total_shares == new_total_shares {
+                // time delta wasn't large enough to pay any management/protocol fee
                 skip_ts_update = true;
             }
 
             management_fee_shares = new_total_shares
                 .cast::<i128>()?
-                .safe_sub(self.total_shares.cast()?)?;
-            self.total_shares = new_total_shares;
-            self.manager_total_fee = self
+                .safe_sub(vault.total_shares.cast()?)?;
+            vault.total_shares = new_total_shares;
+            vault.manager_total_fee = vault
                 .manager_total_fee
                 .saturating_add(management_fee_payment.cast()?);
 
             // in case total_shares is pushed to level that warrants a rebase
-            self.apply_rebase(vault_equity)?;
+            vault.apply_rebase(&mut None, vault_equity)?;
+            Ok(())
+        };
+
+        match vault_protocol {
+            None => {
+                if self.management_fee != 0 && depositor_equity > 0 {
+                    handle_no_protocol_fee(self)?;
+                }
+            }
+            Some(vp) => {
+                if self.management_fee != 0 && vp.protocol_fee != 0 && depositor_equity > 0 {
+                    let since_last = now.safe_sub(self.last_fee_update_ts)?;
+                    let total_fee = self
+                        .management_fee
+                        .safe_add(vp.protocol_fee.cast()?)?
+                        .cast::<i128>()?;
+
+                    // if protocol fee is non-zero and total fee would lead to zero equity remaining,
+                    // so tax equity - 1 but only for the protocol, so that the user is left with 1 and the manager retains their full fee.
+                    let total_fee_payment = depositor_equity
+                        .safe_mul(total_fee)?
+                        .safe_div(PERCENTAGE_PRECISION_I128)?
+                        .safe_mul(since_last.cast()?)?
+                        .safe_div(ONE_YEAR.cast()?)?;
+                    let management_fee_payment = total_fee_payment
+                        .safe_mul(self.management_fee.cast()?)?
+                        .safe_div(total_fee)?;
+                    let protocol_fee_payment = total_fee_payment
```

Comment:
is it safer to just do total_fee_payment - management_fee_payment?

Reply:
didn't think to do that, but yes that works too.